lib/propolis/src/vmm/mem.rs

```diff
@@ -767,23 +789,92 @@ impl<'a, T: AsRef<[SubMapping<'a>]>> MappingExt for T {
             ));
         }
 
-        let iov = self
-            .as_ref()
-            .iter()
-            .map(|mapping| iovec {
-                iov_base: mapping.ptr.as_ptr() as *mut libc::c_void,
-                iov_len: mapping.len,
-            })
-            .collect::<Vec<_>>();
-
-        let read = unsafe {
-            libc::preadv(fd, iov.as_ptr(), iov.len() as libc::c_int, offset)
-        };
-        if read == -1 {
-            return Err(Error::last_os_error());
+        let mut total_capacity = 0;
+        for mapping in self.as_ref().iter() {
+            total_capacity += mapping.len;
         }
 
-        Ok(read as usize)
+        // Gross hack: see the comment on `MAPPING_IO_LIMIT_BYTES`.
+        if total_capacity < MAPPING_IO_LIMIT_BYTES {
```

Comment:
i note that if we are doing a write of _exactly_ 16 MiB we shall spill over into the normal thing. might we want to use `<=` here instead given that the guest is likely to be doing megabyte-aligned-y IOs?